futures-util/src/stream/copy_into.rs

```diff
@@ -0,0 +1,81 @@
+use futures_core::future::Future;
+use futures_core::task::{Context, Poll};
+use futures_core::Stream;
+use futures_io::AsyncWrite;
+use std::io;
+use std::pin::Pin;
+
+/// Future for the [`copy_into`](super::StreamExt::copy_into) method.
+#[derive(Debug)]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct CopyInto<'a, S: Unpin, W: ?Sized + Unpin> {
+    stream: S,
+    read_done: bool,
+    writer: &'a mut W,
+    pos: usize,
+    cap: usize,
+    amt: u64,
+    buf: Option<Vec<u8>>,
```

Comment:
Not sure exactly where you got stuck trying to make this `AsRef<[u8]>` instead, but I'd expect the type definition to look like this:

```rust
pub struct CopyInto<'a, S, W: ?Sized>
where
    S: TryStream,
    S::Ok: AsRef<[u8]>,
{
    stream: Option<S>,
    writer: &'a mut W,
    pos: usize,
    cap: usize,
    amt: u64,
    buf: Option<S::Ok>
}
```